resolwe/flow/utils/docs/autoprocess.py

```diff
@@ -0,0 +1,395 @@
+
+This module introduces a new directive that enables user to document a
+set of processes. The directive can be used in three variations::
+
+    .. autoprocess::
+
+    .. autoprocess:: category_name
+
+    .. autoprocess:: process_slug
+
+The first option documents *all* processes.
+
+The second option documents just the ones that have *category* equal to
+`category_name`. Since some categories have subcategories (analyses,
+analyses:alignment, analyses:variants), this means that subcategories
+will *not* be documented.
+
+Third option decuments only one process: the one with *slug* equal to
+`process_slug`.
+"""
+
+import os
+import re
+import fnmatch
+
+import six
+import yaml
+
+from docutils import nodes
+from docutils.parsers.rst import Directive
+
+
+class AutoProcessDirective(Directive):
+    """
+    Directive used to automatically document Resolwe process.
+    """
+    has_content = True
+    required_arguments = 0
+    optional_arguments = 0
+    final_argument_whitespace = False
+    option_spec = None
+
+    def filter_processes(self, content, categories, processes_dict):
+        """
+        Filter processes list by the content of the directive
+
+        Content can be:
+
+            * None - if there is not content, all processes are returned.
+            * Category - if the content of the directive is a string matching
+              an element in categories list, only processes with this
+              category will be returned.
+            * Process slug - if the content of directive is a valid process
+              slug only corresponding process is returned. Actually, a list
+              with only this process is returned.
+            * If none of this options, then raise a ValueError
+
+        :param list content: content of the directive
+        :param set categories: set of all categories
+        :param dict processes_dict: dict of all source_base_url: processes pairs
+        :rtype: list
+        """
+        if not content:
+            return processes_dict
+        elif len(content) > 1:
+            raise NotImplementedError("Only one argument can be passes to autoprocess.")
+        content = content[0]
+        # Make a filtered dict of all process that have name == content
+        querry_for_slug = dict(pair for pair in processes_dict.items() if pair[1]['slug'] == content)
+        if len(querry_for_slug) == 1:
+            return querry_for_slug
+
+        elif content in categories:
+            filtered_dict = dict(
+                pair for pair in processes_dict.items()
+                if pair[1].get('category', 'uncategorized') == content)
+            return filtered_dict
+
+        raise ValueError("Invalid argument given to autoprocess directive.")
+
+    def get_process_definition_start(self, fname, slug):
+        """
+        Find the starting line of process definition - where slug is defined
+
+        :param path/str fname: Filename with processes
```

Comment:
I think `path/str` type definition is not accepted by Sphinx. Just use `str` since that is the type of the parameter and perhaps change the description to something like "Path of the file where process is defined"?